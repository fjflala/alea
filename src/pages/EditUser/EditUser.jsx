/**
 * Module dependencies
 */
import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

/**
 * Actions
 */
import { usersGetUser, usersEditUser, usersCleanEditedUser } from '../../modules/Users/actions';

/**
 * UI Components
 */
import Card from '../../components/Card';
import Button from '../../components/Button';
import Input from '../../components/Input';
import Header from '../../components/Header';

/**
 * Import styles
 */
import './styles.scss';

/**
 * Represents a view of Edit User
 */
export const EditUser = ({
  match: {
    params: {
      userID,
    },
  },
  selectedUser,
  editedUser,
  getUser,
  editUser,
  cleanEditedUser,
}) => {
  const [user, setUser] = useState({
    avatar: '',
    first_name: '',
    last_name: '',
    email: '',
  });

  useEffect(() => {
    getUser({ userID });
    return () => {
      cleanEditedUser();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (selectedUser) {
      setUser(selectedUser);
    }
  }, [selectedUser]);

  const onChangeValue = (field, value) => {
    setUser({
      ...user,
      [field]: value,
    });
  };

  const onSubmit = (event) => {
    event.preventDefault();

    editUser({
      userID: user.id,
      data: user,
    });
  };

  return (
    <React.Fragment>
      <Header />
      <section className="edit-user">
        <header>
          <h1>Edit User Page</h1>
        </header>
        <Card className="edit-user__card">
          {Object.keys(editedUser).length > 0 && (
            <p className="edit-user__message--success">
              Your user has been updated successfully.
              <br />
              <Link to="/users">Go to users list</Link>
            </p>
          )}
          <form onSubmit={onSubmit}>
            <img src={user.avatar} alt={user.first_name} />
            <br />
            <Input
              type="text"
              placeholder="Email"
              value={user.email}
              onChange={e => onChangeValue('email', e.target.value)}
            />
            <br />
            <Input
              type="text"
              placeholder="First Name"
              value={user.first_name}
              onChange={e => onChangeValue('first_name', e.target.value)}
            />
            <br />
            <Input
              type="text"
              placeholder="Last Name"
              value={user.last_name}
              onChange={e => onChangeValue('last_name', e.target.value)}
            />
            <br />
            <Link to="/users" className="ui-button ui-button--secondary edit-user__cancel-button">
              Cancel
            </Link>
            <Button type="submit" modifier="primary" className="edit-user__save-button">Save</Button>
          </form>
        </Card>
      </section>
    </React.Fragment>
  );
};

/**
 * PropTypes definitions 
 */
EditUser.propTypes = {
  match: PropTypes.shape({}).isRequired,
  selectedUser: PropTypes.shape({}),
  editedUser: PropTypes.shape({}),
  getUser: PropTypes.func.isRequired,
  editUser: PropTypes.func.isRequired,
  cleanEditedUser: PropTypes.func.isRequired,
};

EditUser.defaultProps = {
  selectedUser: {},
  editedUser: {},
}

/**
 * Connect to store
 */
const mapStateToProps = state => ({
  selectedUser: state.users.selectedUser,
  editedUser: state.users.editedUser,
});

const mapDispatchToProps = {
  getUser: usersGetUser,
  editUser: usersEditUser,
  cleanEditedUser: usersCleanEditedUser,
};

/**
 * Expose component
 */
export default connect(mapStateToProps, mapDispatchToProps)(EditUser);
