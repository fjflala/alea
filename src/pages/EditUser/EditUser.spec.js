import React from 'react';
import renderer from 'react-test-renderer';
import { EditUser } from './EditUser';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { mount } from 'enzyme';

describe('<EditUser />', () => {
  let props;
  beforeEach(() => {
    props = {
      match: {
        params: {
          userID: 1,
        },
      },
      selectedUser: {
        id: 1,
        first_name: 'Dan',
        last_name: 'Abramov',
        email: 'dan@abramov.net',
      },
      getUser: jest.fn(),
      editUser: jest.fn(),
      cleanEditedUser: jest.fn(),
    };
  });

  it('should correctly render', () => {
    const tree = renderer
      .create(<Router history={createMemoryHistory({route: ['/user/1/edit']})}>
        <EditUser {...props} />
      </Router>)
      .toJSON();
    
    expect(tree).toMatchSnapshot();
  });

  it('should call register method when click on register button', () => {
    const wrapper = mount(<Router history={createMemoryHistory({route: ['/user/1/edit']})}>
      <EditUser {...props} />
    </Router>);
    
    expect(props.getUser).toHaveBeenCalled();

    wrapper.find('form').simulate('submit');
    
    expect(props.editUser).toHaveBeenCalled();
  });

  it('should show an error message', () => {
    props.editedUser = {
      id: 1,
      first_name: 'Dan',
      last_name: 'Abramov',
      email: 'dan@abramov.net',
      updated_at: '09/12/2018',
    };
    
    const wrapper = mount(<Router history={createMemoryHistory({route: ['/user/1/edit']})}>
      <EditUser {...props} />
    </Router>);
    
    expect(wrapper.find('.edit-user__message--success').length).toBe(1);
  });
});