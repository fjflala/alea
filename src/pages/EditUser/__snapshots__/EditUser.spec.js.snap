// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditUser /> should correctly render 1`] = `
Array [
  <header
    className="ui-header"
  >
    <figure>
      <img
        alt="Alea"
        src="https://www.alea.com/wp-content/themes/alea/library/images/logo-alea.svg"
      />
    </figure>
  </header>,
  <section
    className="edit-user"
  >
    <header>
      <h1>
        Edit User Page
      </h1>
    </header>
    <div
      className="ui-card edit-user__card"
    >
      <form
        onSubmit={[Function]}
      >
        <img
          alt=""
          src=""
        />
        <br />
        <input
          className="ui-input "
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Email"
          type="text"
          value=""
        />
        <br />
        <input
          className="ui-input "
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="First Name"
          type="text"
          value=""
        />
        <br />
        <input
          className="ui-input "
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Last Name"
          type="text"
          value=""
        />
        <br />
        <a
          className="ui-button ui-button--secondary edit-user__cancel-button"
          href="/users"
          onClick={[Function]}
        >
          Cancel
        </a>
        <button
          className="ui-button ui-button--primary edit-user__save-button"
          onClick={[Function]}
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </section>,
]
`;
